salt/utils/jinja.py

```diff
@@ -206,6 +211,303 @@ def ensure_sequence_filter(data):
     return data
 
 
+def to_bool(val):
+    '''
+    Returns the logical value.
+
+    .. code-block:: jinja
+
+        {{ 'yes' | to_bool }}
+
+    will be rendered as:
+
+    .. code-block:: text
+
+        True
+    '''
+    if val is None:
+        return False
+    if isinstance(val, bool):
+        return val
+    if isinstance(val, (six.text_type, six.string_types)):
+        return val.lower() in ('yes', '1', 'true')
```

Comment:
This is a departure from the pythonic interpretation of a truthy string and seems wrong to me:
```py
>>> bool('False')
True
>>> bool('')
False
```
But then again jinja is not a real language so should I really be expecting consistency?

Reply:
I think not (val.lower() in ('no', '0', 'false')) is a more correct return value.

Reply:
> This is a departure from the pythonic interpretation of a truthy string and seems wrong to me:

I believe this is present to match how YAML treats these values, not Jinja.

> I think `not (val.lower() in ('no', '0', 'false'))` is a more correct return value.

I'm failing to understand the distinction. Just seems more complicated.

Reply:
The distinction is that in changes `{{ 'a string' | to_bool }}` from `false` to `true`.  According to my reading of the yaml spec, boolean types are limited to a small set of specific values, meaning that the truth value of `'a string'` is undefined.  My argument is in that case to use the pythonic convention rather than the opposite.

Reply:
That's intended to be caught just below...

```python
    if not isinstance(val, collections.Hashable):
        return len(val) > 0
```

Reply:
> I believe this is present to match how YAML treats these values, not Jinja.
Also this is a jinja filter and so being available inside purely jinja constructs will lead to contradictory results:
```jinja
{% set result = 'a string' | to_bool %}
{% if result %}
  {% set var = 'all ye need to know' %}
{% endif %}
```

Reply:
hm, perhaps it could be caught elsewhere, however the
```py
   if isinstance(val, (six.text_type, six.string_types)):
```
will unconditionally capture all strings.
```py
$ python
Python 3.8.5 (default, May 27 2021, 13:30:53) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import salt.utils.jinja
>>> salt.utils.jinja.to_bool('a string')
False
>>> 
```

Reply:
Ahh, I see now. Yeah, a test case for that would be great. :)